File: AbstractMethod/python/Ex3.py
# PALETA DE CORES
class UI:
    reset = "\033[0m"
    bold = "\033[1m"
    pink = "\033[95m"
    blue = "\033[94m"
    purple = "\033[38;5;54m"
    white = "\033[97m"

ui = UI()

def header(titulo, color):
    print("\n" + color + ui.bold + "╔══════════════════════════════════╗")
    print(color + f"║  {titulo.ljust(30)} ║")
    print("╚══════════════════════════════════╝" + ui.reset)

def divider(color):
    print(color + " ────────────────────────────────" + ui.reset)

# PRODUTOS ABSTRATOS
class Botao:
    def render(self):
        raise NotImplementedError("render() deve ser implementado")

class Janela:
    def render(self):
        raise NotImplementedError("render() deve ser implementado")

# PRODUTO CONCRETO - WINDOWS
class BotaoWindows(Botao):
    def render(self):
        return "🟪 Botão estilo Windows"

class JanelaWindows(Janela):
    def render(self):
        return "🪟 Janela estilo Windows"

# PRODUTO CONCRETO - MAC
class BotaoMac(Botao):
    def render(self):
        return "⬜ Botão estilo MacOS"

class JanelaMac(Janela):
    def render(self):
        return "🖥️ Janela estilo MacOS"

# PRODUTO CONCRETO - LINUX
class BotaoLinux(Botao):
    def render(self):
        return "🟩 Botão estilo Linux"

class JanelaLinux(Janela):
    def render(self):
        return "🖥️ Janela estilo Linux"

# FÁBRICA ABSTRATA
class FabricaGUI:
    def criar_botao(self):
        raise NotImplementedError("criar_botao() deve ser implementado")

    def criar_janela(self):
        raise NotImplementedError("criar_janela() deve ser implementado")

# FÁBRICAS CONCRETAS
class FabricaWindows(FabricaGUI):
    def criar_botao(self):
        return BotaoWindows()

    def criar_janela(self):
        return JanelaWindows()

class FabricaMac(FabricaGUI):
    def criar_botao(self):
        return BotaoMac()

    def criar_janela(self):
        return JanelaMac()

class FabricaLinux(FabricaGUI):
    def criar_botao(self):
        return BotaoLinux()

    def criar_janela(self):
        return JanelaLinux()

# CLIENTE
def iniciar_sistema(fabrica, color):
    botao = fabrica.criar_botao()
    janela = fabrica.criar_janela()

    print(ui.white + "   ➤ " + botao.render() + ui.reset)
    print(ui.white + "   ➤ " + janela.render() + ui.reset)
    divider(color)

# USO
header("💻 SISTEMA WINDOWS", ui.pink)
iniciar_sistema(FabricaWindows(), ui.pink)

header("🍎 SISTEMA MAC", ui.blue)
iniciar_sistema(FabricaMac(), ui.blue)

header("🐧 SISTEMA LINUX", ui.purple)
iniciar_sistema(FabricaLinux(), ui.purple)

File: AbstractMethod/JavaScript/Ex3.js
// PALETA DE CORES
const ui = {
    reset: "\x1b[0m",
    bold: "\x1b[1m",
    pink: "\x1b[95m",     
    blue: "\x1b[94m",     
    purple: "\x1b[38;5;54m",   
    white: "\x1b[97m"
};

function header(titulo, color) {
    console.log("\n" + color + ui.bold + "╔══════════════════════════════════╗");
    console.log(color + `║  ${titulo.padEnd(30)} ║`);
    console.log("╚══════════════════════════════════╝" + ui.reset);
}

function divider(color) {
    console.log(color + "   ────────────────────────────────" + ui.reset);
}

// PRODUTOS ABSTRATOS
class Botao {
    render() { throw new Error("render() deve ser implementado"); }
}

class Janela {
    render() { throw new Error("render() deve ser implementado"); }
}

// PRODUTO CONCRETO -  WINDOWS
class BotaoWindows extends Botao {
    render() { return "🟪 Botão estilo Windows"; }
}

class JanelaWindows extends Janela {
    render() { return "🪟 Janela estilo Windows"; }
}

// PRODUTO CONCRETO -  MAC
class BotaoMac extends Botao {
    render() { return "⬜ Botão estilo MacOS"; }
}

class JanelaMac extends Janela {
    render() { return "🖥️ Janela estilo MacOS"; }
}

// PRODUTO CONCRETO - LINUX
class BotaoLinux extends Botao {
    render() { return "🟩 Botão estilo Linux"; }
}

class JanelaLinux extends Janela {
    render() { return "🖥️ Janela estilo Linux"; }
}

// FÁBRICA ABSTRATA
class FabricaGUI {
    criarBotao() { throw new Error("criarBotao() deve ser implementado"); }
    criarJanela() { throw new Error("criarJanela() deve ser implementado"); }
}

// FÁBRICAS CONCRETAS
class FabricaWindows extends FabricaGUI {
    criarBotao() { return new BotaoWindows(); }
    criarJanela() { return new JanelaWindows(); }
}

class FabricaMac extends FabricaGUI {
    criarBotao() { return new BotaoMac(); }
    criarJanela() { return new JanelaMac(); }
}

class FabricaLinux extends FabricaGUI {
    criarBotao() { return new BotaoLinux(); }
    criarJanela() { return new JanelaLinux(); }
}

// CLIENTE
function iniciarSistema(fabrica, color) {
    const botao = fabrica.criarBotao();
    const janela = fabrica.criarJanela();

    console.log(ui.white + "   ➤ " + botao.render() + ui.reset);
    console.log(ui.white + "   ➤ " + janela.render() + ui.reset);
    divider(color);
}

// USO
header("💻 SISTEMA WINDOWS", ui.pink);
iniciarSistema(new FabricaWindows(), ui.pink);

header("🍎 SISTEMA MAC", ui.blue);
iniciarSistema(new FabricaMac(), ui.blue);

header("🐧 SISTEMA LINUX", ui.purple);
iniciarSistema(new FabricaLinux(), ui.purple);